File: basket.go
// a struct must represent the @product hash with
// the properties: 'name', 'flavor' and 'kind'

// another struct named basket will be an array of @products
// we will define a function and attach it to basket to interact with product


package main

import "fmt"

type basket []product

type product struct {
  name string
  flavour string
  kind string
}


func (p *basket) add_item(entry product){
  *p = append(*p, entry)

  fmt.Printf("Entry %s created!\n", entry.name)
}

func (p basket) change_item(name string, entry product) {
  for key, val := range p {
    if val.name ==  name {
      p[key] = entry
    }
  }

  fmt.Printf("Item %s changed!\n", name)
}

func (p basket) items(){
  fmt.Printf("There are %d number of items in the basket\n", len(p))

  for _,val := range p {
    fmt.Printf("Name: %s\n" +
               "Flavour: %s\n" +
               "Kind: %s\n",
               val.name,
               val.flavour,
               val.kind)
    fmt.Printf("\n")
  }
}


func main(){
  basket := new(basket)

  basket.add_item(
    product{
      name: "apple",
      flavour: "its juicy",
      kind: "fruit",
    },
  )

  basket.add_item(
    product{
      name: "carrot",
      flavour: "",
      kind: "veggies",
    },
  )

  basket.change_item("carrot",
                     product{
                       name: "cucumber",
                       flavour: "Slightly bitter with a mild melon aroma",
                       kind: "veggies",
                     },
                   )

  basket.items()
}

File: basket.rb
require 'pry'

class Basket
  def initialize
    @produce = []
  end

  def add_item(**entry)
    # Raise an error if name, flavour and kind
    # keys are not passed

    items = %w[name flavour kind]

    unless items.any? { |key| entry.key? key.to_sym}
      raise "Usage: add_item(name: '...',
                             flavour: '...',
                             kind: '....'
                      )"
    end

    @produce.push(entry)

    puts "Entry #{entry[:name]} created!"
  end

  def change_item(name, entry)
    # Deleting existing record

    item = @produce
           .delete_if { |h| h[:name] == name}
           .first
           .dup

    item = entry

    # Add it to the instance variable
    @produce.push(item)
    @produce.uniq!

    puts "Item #{name} changed!"
  end

  def items
    puts "There are #{@produce.count} number of items in the basket"

    @produce.each do |entry|
      item(entry)
    end
  end

  private

  def item(entry)
    puts "Name: #{entry[:name]}"
    puts "Flavour: #{entry[:flavour]}"
    puts "Kind: #{entry[:kind]}"
  end
end

basket = Basket.new
basket.add_item(
  name: 'apple',
  flavour: "It's a little sour and bitter, but mostly sweet, very juicy",
  kind: 'fruit'
)

basket.add_item(
  name: 'carrot',
  flavour: '',
  kind: 'veggies'
)

basket.change_item(
  'carrot',
  name: 'cucumber',
  flavour: 'slightly bitter with a mild melon aroma and planty flavor',
  kind: 'veggies'
)


basket.items
